src/branch_and_prune.jl

```diff
@@ -5,42 +5,98 @@ export branch_and_prune, Bisection, Newton
 
 diam(x::Root) = diam(x.interval)
 
-
 Base.size(x::Interval) = (1,)
 
 isinterior{N}(X::IntervalBox{N}, Y::IntervalBox{N}) = all(isinterior.(X, Y))
 
+# contractors:
+"""
+    Contractor{F}
 
+    Abstract type for contractors.
 """
-    branch_and_prune(f, X, contractor, tol=1e-3)
+abstract type Contractor{F} end
 
-Generic branch and prune routine for finding isolated roots of a function
-`f:R^n → R^n` in a box.
+export Bisection, Newton
 
-Inputs:
-- `f`: function whose roots will be found
-- `X`: `Interval` or `IntervalBox`
-- `contractor`: function that, when applied to the function `f`, determines
-    the status of a given box `X`. It returns the new box and a symbol indicating
-    the status. Current possible values are `Bisection` and `Newton`.
+struct Bisection{F} <: Contractor{F}
+    f::F
+end
 
-"""
-function branch_and_prune(f, X, contractor, tol=1e-3)
+Bisection{n}(::Type{Val{n}}, f) = Bisection(f)
+
+function (contractor::Bisection)(X)
+    image = contractor.f(X)
+
+    if !(contains_zero(image))
+        return :empty, X
+    end
+
+    return :unknown, X
+end
+
+
+struct Newton{F,FP,O} <: Contractor{F}
+    f::F
+    fp::FP
+    op::O
+end
+
+function Newton(::Type{Val{1}}, f::Function)
+    f_prime = x -> ForwardDiff.derivative(f, x)
+    Newton(f, f_prime, N)
+end
+
+function Newton{n}(::Type{Val{n}}, f::Function)
+    f_prime = x -> ForwardDiff.jacobian(f, x)
+    Newton(f, f_prime, N)
+end
+
+function Newton(f_prime::Function)
+    return NewtonConstructor(f_prime)
+end
+
+function (C::Newton)(X)
+
+    # use Bisection contractor for this:
+    if !(contains_zero(IntervalBox(C.f(X))))
+        return :empty, X
+    end
+
+    # given that have the Jacobian, can also do mean value form
+
+    NX = C.op(C.f, C.fp, X) ∩ X
+
+    isempty(NX) && return :empty, X
+
+    if NX ⪽ X  # isinterior; know there's a unique root inside
+        NX =  refine(X -> C.op(C.f, C.fp, X), NX)
+        return :unique, NX
+    end
+
+    return :unknown, NX
+end
 
-    input_dim = length(X)
-    output_dim = length(X)
 
-    # @show input_dim
-    # @show output_dim
+struct NewtonConstructor{FP}
+    f_prime::FP
+end
+
 
-    # if !(input_dim == output_dim)
-    #     throw(ArgumentError("Input dimension ($input_dim) and output dimension ($output_dim) must be the same."))
-    # end
+"""
+    branch_and_prune(X, contract, tol=1e-3)
```

Comment:
Can this function just be renamed `roots` now?

Reply:
It could as no other `roots` function conflict with it. However from a design point of view separating the function actually containing the algorithm from the setup function (in our case `roots`) may add readability.

On the other hand if it is named `roots` it will be exported, allowing for syntax like

    roots(x -> x^2 - 2, Newton(f, f_prime), -10..10